torchrl/modules/models/multiagent.py

```diff
@@ -158,22 +163,45 @@ def __init__(
         self.share_params = share_params
         self.centralised = centralised
 
-        self.agent_networks = nn.ModuleList(
-            [
-                MLP(
-                    in_features=n_agent_inputs
-                    if not centralised
-                    else n_agent_inputs * n_agents,
-                    out_features=n_agent_outputs,
-                    depth=depth,
-                    num_cells=num_cells,
-                    activation_class=activation_class,
-                    device=device,
-                    **kwargs,
-                )
-                for _ in range(self.n_agents if not self.share_params else 1)
-            ]
-        )
+        def make_net():
+            return MLP(
+                in_features=n_agent_inputs
+                if not centralised
+                else n_agent_inputs * n_agents,
+                out_features=n_agent_outputs,
+                depth=depth,
+                num_cells=num_cells,
+                activation_class=activation_class,
+                device=device,
+                **kwargs,
+            )
+
+        if not self.share_params:
+            agent_networks = [make_net() for _ in range(self.n_agents)]
+            self.params = TensorDictParams(
+                torch.stack(
+                    [TensorDict.from_module(mod) for mod in agent_networks], 0
+                ).contiguous(),
+                no_convert=True,
+            )
+            net = agent_networks[0]
+        else:
+            net = make_net()
+            self.params = TensorDictParams(TensorDict.from_module(net), no_convert=True)
+        make_functional(net)
+        self.net = net
+        if self.centralised:
+            if not self.share_params:
+                self.net_call = torch.vmap(self.net, in_dims=(None, 0), out_dims=(-2,))
+            else:
+                self.net_call = self.net
+        else:
+            if not self.share_params:
+                self.net_call = torch.vmap(self.net, in_dims=(-2, 0), out_dims=(-2,))
+            else:
+                self.net_call = torch.vmap(self.net, in_dims=(-2, None), out_dims=(-2,))
```

Comment:
is this neded? wouldn't the mlp do vectorization by default?

Reply:
Not sure all ops will support an arbitrary num of dims

Reply:
oh ok let's revert to previous version then and we are good to go